Parser Utility - Linked List and Environment

Hikari9 opened the pull request:
# Summary

Propose for the two utility classes (LinkedList and Environment) to be used for variable scoping during compilation of Rainscript.

## Persistent Linked List

The linked lists here are persistent, hence is most similar to racket language's `cons`. This will be used to persistently log warnings/errors during recursive descent.

## Persistent Environment

The Environment will be similar to the linked list, however keeps both a key and a value instead of a single data head. Though the two are similar, they will have no superclass/subclass relationship since they handle their methods differently. The Environment will mainly be used for variable identifier lookups and block-scoping in Rainscript.

The `push` method cheaply creates a new Environment every time. This style is used so that it is easier to `reset` block scopes during recursive-descent in, for example, `#if` or `#choice` blocks.

> Note: Environment is NOT for variable content but for data types, since this is just the compiler.

# How to Test

1. Have a fresh clone of this PR by running:
```
git clone https://github.com/Hikari9/rainscript -b parser-00-util rainscript-parser-00-util
cd rainscript-parser-00-util
```
  Alternatively, you may also fetch using a remote to `Hikari9/parser-00-util`.

2. For first use, make sure npm is installed then run:
```
npm install
```

3. To perform tests on your machine, run:
```
npm test
```

4. (Optional) Add your own tests to `/tests/parser/util/LinkedList.js` or `/tests/parser/util/Environment.js`. Push the new tests into this PR so that it can be reviewed (if you are a maintainer). Otherwise simply leave a comment.

Hikari9 committed:
Merge branch 'travis' into parser-00-util

Hikari9 committed:
Add Babel ES6 and mocha dependencies to package.json

Commands 'npm test' and 'npm run fix' should now work without errors with this commit.

Hikari9 committed:
[parser] Add LinkedList class

Hikari9 committed:
[parser] Add Environment-type linked list

Hikari9 committed:
[parser] Add tests for Environment

Hikari9 committed:
[parser] Prefer strictEqual assertions

Vaan52 commented on tests/parser-util-LinkedList.js:
```diff
@@ -0,0 +1,54 @@
+          'list 3 parent not common');
+        assert.strictEqual(list4.head(), 'world', 'list 4 head world');
+        assert.strictEqual(list4.tail().head(), 'not common',
+          'list 3 parent is missing');
```
What do you mean by `list 3 parent is missing`?

Hikari9 replied:
Supposedly the assertion checks if `list3` is still the parent of `list4`, but this test is already deprecated as of the most recent commit :)

Vaan52 commented on tests/parser-util-LinkedList.js:
```diff
@@ -0,0 +1,54 @@
+          'list 4 parent common');
+        // set new head
+        list2.head('not common');
+        assert.strictEqual(list3.head(), 'hello', 'list 3 head common');
```
Although the test is correct, the name should be `list 3 head hello`

Hikari9 replied:
Already fixed this with new `should` syntax in the latest commit.

Hikari9 committed:
Merge pull request (#4) "Add .travis.yml"

* Add .travis.yml

* Add Babel ES6 and mocha dependencies to package.json

Commands 'npm test' and 'npm run fix' should now work without errors with this commit.

* [babel-es6] Do not ignore package-lock.json

* [travis] Add before_script and script modules

erocrizs commented on tests/parser-util-Environment.js:
```diff
@@ -0,0 +1,48 @@
+      });
+    });
+  });
+});
```
move/rename file to be `tests/parser/util/Environment.js`

Hikari9 replied:
Alright, will restructure the `tests/` folder accordingly :D

erocrizs commented on tests/parser-util-LinkedList.js:
```diff
@@ -0,0 +1,54 @@
+      });
+    });
+  });
+});
```
move/rename file to be `tests/parser/util/LinkedList.js`

erocrizs commented on tests/parser-util-LinkedList.js:
```diff
@@ -0,0 +1,54 @@
+import { LinkedList } from '../parser/util/LinkedList';
+const assert = require('assert');
+
```
Can we use `should` module of chai instead? It's a little bit easier to read.

erocrizs commented on tests/parser-util-LinkedList.js:
```diff
@@ -0,0 +1,54 @@
+        assert.strictEqual(list.head(), undefined);
+        assert.strictEqual(list.tail(), undefined);
+      });
+      it('should have correct links', () => {
```
Can we separate each it-block with empty lines so it's easier to visually group?

Hikari9 replied:
Makes sense. Will add empty lines between `it` statements to make them more readable.

erocrizs commented on tests/parser-util-Environment.js:
```diff
@@ -0,0 +1,48 @@
+        }
+        assert.strictEqual(list.lookup('one'), 1, 'one');
+        assert.strictEqual(list.lookup('two'), 2, 'one');
+        assert.strictEqual(list.lookup('three'), 3, 'one');
```
Why is the assert description `one`?

Hikari9 replied:
Forgot to remove unnecessary `one`s, my bad

erocrizs commented on parser/util/LinkedList.js:
```diff
@@ -0,0 +1,23 @@
+   */
+  constructor (head, tail) {
+    let size = tail instanceof LinkedList ? tail.size() + 1 : 0;
+    this.head = (h) => h === undefined ? head : (head = h);
```
A couple of comments:
1. Let's use package for checking for undefines `_.isUndefined(...)` for better readability
2. Restructure this lambda function to be more easily readable:
```javascript
(h) => {
  if (! _.isUndefined(h)) {
    head = h;
  }
  return head;
}
```

Hikari9 replied:
Will change this to a more readable lambda expression.

erocrizs commented on parser/util/Environment.js:
```diff
@@ -0,0 +1,35 @@
+export class Environment extends LinkedList {
+  constructor (head, tail) {
+    super(head, tail);
+    let hed = this.head;
```
Use `const`

Also, please use a different name. The name `hed` is not very descriptive.

erocrizs commented on parser/util/Environment.js:
```diff
@@ -0,0 +1,35 @@
+  constructor (head, tail) {
+    super(head, tail);
+    let hed = this.head;
+    this.head = (value) => value === undefined ? hed() : hed([hed()[0], value]);
```
Refactor to be more readable. Same with my comment with the `LinkedList.head()` details.

erocrizs commented on parser/util/Environment.js:
```diff
@@ -0,0 +1,35 @@
+   * @param {*} value the value associated with the key
+   * @return a new Environment with the appended key-value pair
+   */
+  push (key, value) {
```
1. What will happen if I do `env.push('abc')`? Will it just create a variable with an `undefined` value? Is that the intended usage?
2. Also, what will happen if I do `env.push()`?

Hikari9 replied:
Noted. Undefined value should be allowed, but undefined key should not (should raise exception). Will add arity checking in the next commit.

Hikari9 replied:
Changed my mind, `undefined` value should not be allowed as well, otherwise the `Environment.lookup` method returning `undefined` will make no sense.

erocrizs reviewed (changes requested):
A few comments. Please respond. Thanks!

Hikari9 reviewed (commented):
Will address these issues in the next commit :)

Hikari9 committed:
[travis] Send travis CI notifications to slack (#7)

Sends notifications to team slack channel. Also fixed minor spelling mistake.

Hikari9 committed:
Add Babel ES6 and mocha dependencies to package.json (#5)

* Add Babel ES6 and mocha dependencies to package.json

Commands 'npm test' and 'npm run fix' should now work without errors with this commit.

* [babel-es6] Do not ignore package-lock.json

* [babel-es6] Cleanup

Hikari9 committed:
Merge remote-tracking branch 'upstream/master' into parser-00-util

Hikari9 committed:
[parser] [fix] Restructure test files

Mirror folder structure of parser/ into tests/parser/ so that each script has a direct mapping to its test file.

Hikari9 committed:
[parser] [fix] Make LinkedList and Environment more persistent

Fix arity checking of methods, prefer getters without setters to keep everything persistent, improve tests

Hikari9 committed:
[parser] [fix] Add more tests for error checking

Hikari9 committed:
[parser] [fix] Adopt Chai should syntax for tests

Hikari9 commented:
@erocrizs @Vaan52 Latest set of commits contains the following changes:

1. Restructure `tests/` folder to mirror folder structure of `parser/` folder
    - minor changes to `npm test` script to include subfolders
2. Change `x === undefined` to `_.isUndefined(x)`
3. Tests now follow `chai.should` syntax for readability
    - minor change to Eslint configuration file to follow this change
4. LinkedList does not allow setting head and tail anymore
    - it was intended to be persistent in the first place, so setters shouldn't exist
    - `head`, `tail`, and `size` are now made into getter functions in the class prototype
    - added additional tests for this change as well

Hikari9 committed:
[parser] [fix] Add chai-friendly plugin to eslint

Ignores no-unused-expressions error during linting when using chai.expect or chai.should

Hikari9 committed:
[parser] [fix] Add more tests for error checking

Also add Doc Block comments for each function that throws an Error.

erocrizs commented on parser/util/Environment.js:
```diff
@@ -0,0 +1,78 @@
+   * @throws {Error} when key or value is undefined
+   * @return a new Environment with the appended key-value pair
+   */
+  push (key, value) {
```
Since `push` no longer supports single-parameter calls, hence an interface of the superclass is not supported by the subclass, I don't see any more reason to keep `Environment` as a subclass for `LinkedList`. This violates Liskov's substitution principle.

While this does not matter very much in a dynamic language like JS, I think we should change `Environment` to be its own independent class as good practice, and so that people who will use this for TS projects will not make the mistake of assigning an Environment instance to a LinkedList variable and calling a single-parameter `push`.

What do you think?

Hikari9 replied:
Your reasoning makes sense to me. I agree that it's better to separate them into their own classes, since their methods diverge already albeit having the same method names. Will do isolate Environment as its own class. Will add that in the next commit.

However, I would like to keep the persistent LinkedList since will be using that to list errors/warnings, if that's okay :)

erocrizs replied:
Sure

Hikari9 commented:
Updated the pull request description to match the new semantics of LinkedList and Environment.

Hikari9 committed:
[parser] Don't make Environment extend LinkedList

Subclassing is unnecessary  for same method names with different meanings

erocrizs reviewed (approved):
Edit the description/first comment of this PR to include steps to test the module and have this tested by others.

You may merge this after at least two "Test Pass" comments.

Hikari9 commented:
Added "How to Test" to the PR description. You may add new tests if needed :)

Vaan52 commented:
@Hikari9 Please make eslint be agnostic in its line break style. It's currently in unix. It's found in .eslintrc.json. We Windows users will get errors.

erocrizs commented:
Tested on Ubuntu. Test Passed.

Hikari9 commented:
@Vaan52 thanks, will turn it off in eslint in the next commit.

Hikari9 committed:
[parser] Turn off linebreak-style in eslint

Vaan52 commented:
Tested again on Windows 10. Test Passed.
